tensorflow_addons/layers/crf.py

```diff
@@ -0,0 +1,491 @@
+            initializer=self.chain_initializer,
+            regularizer=self.chain_regularizer,
+            constraint=self.chain_constraint,
+        )
+
+        # bias that works with self.kernel
+        if self.use_kernel and self.use_bias:
+            self.bias = self.add_weight(
+                shape=(self.units, ),
+                name="bias",
+                initializer=self.bias_initializer,
+                regularizer=self.bias_regularizer,
+                constraint=self.bias_constraint,
+            )
+        else:
+            self.bias = 0
+
+        # weight of <START> to tag probability and tag to <END> probability
+        if self.use_boundary:
+            self.left_boundary = self.add_weight(
+                shape=(self.units, ),
+                name="left_boundary",
+                initializer=self.boundary_initializer,
+                regularizer=self.boundary_regularizer,
+                constraint=self.boundary_constraint,
+            )
+            self.right_boundary = self.add_weight(
+                shape=(self.units, ),
+                name="right_boundary",
+                initializer=self.boundary_initializer,
+                regularizer=self.boundary_regularizer,
+                constraint=self.boundary_constraint,
+            )
+
+        # or directly call self.built = True
+        super(CRF, self).build(input_shape)
+
+    def call(self, inputs, mask=None, **kwargs):
+        # mask: Tensor(shape=(batch_size, sequence_length), dtype=bool) or None
+
+        if mask is not None:
+            assert (tf.keras.backend.ndim(mask) == 2
+                    ), "Input mask to CRF must have dim 2 if not None"
+
+        # left padding of mask is not supported, due the underline CRF function
+        # detect it and report it to user
+        first_mask = None
+        if mask is not None:
+            left_boundary_mask = self._compute_mask_left_boundary(mask)
+            first_mask = left_boundary_mask[:, 0]
+
+        # remember this value for later use
+        self.mask = mask
+
+        if first_mask is not None:
+            with tf.control_dependencies([
+                    tf.debugging.assert_equal(
+                        tf.math.reduce_all(first_mask),
+                        tf.constant(True),
+                        message="Currently, CRF layer do not support left padding"
+                    )
+            ]):
+                self.potentials = self._dense_layer(inputs)
+        else:
+            self.potentials = self._dense_layer(inputs)
+
+        # appending boundary probability info
+        if self.use_boundary:
+            self.potentials = self.add_boundary_energy(
+                self.potentials, mask, self.left_boundary, self.right_boundary)
+
+        self.sequence_length = self._get_sequence_length(inputs, mask)
+
+        decoded_sequence, _ = self.get_viterbi_decoding(
+            self.potentials, self.sequence_length)
+
+        return decoded_sequence
```

Comment:
Is it possible to return tf.one_hot here to make it look like logits so that it could be compatible with tf.keras.metrics?

Reply:
As far as I know,  metric functions (for example tf.keras.metrics.Accuracy) can work on the current output format. Are there some metrics that one_hot can do but the current output format can't?

Reply:
What I'm trying to do here is make an example with CRF and SparseCategoricalCrossentropy look the same. So the example metric is SparseCategoricalAccuracy, but in the real life it may be some custom metrics like Precision/Recall on named entities in NER

Reply:
Thank you for your explain, I know what you mean now. Since this is the first release of CRF layer, I am trying to keep it as simple as possible so it can release quickly. What you want is the layer has a friendly output for builtin metrics and custom metric which is an important feature for me too. I am planning to write some custom metrics around the NER when the CRF layer is ready.  But I am not sure one_hot format is good for everyone. Maybe we can try using the CRF after it is released, if we still have trouble with the output format, then we can have a detailed plan how to make it better in the next release of CRF layer.